File: 907-koko-eating-bananas/koko-eating-bananas.java
class Solution {
    private int sum(int[] piles){
        int maxi = 0;
        for(int i: piles){
            maxi = Math.max(i,maxi);
        }
        return maxi;
    }
    private boolean isPossible(int mid, int[] piles, int h){
        int total = 0;
        for(int i: piles){
            total += Math.ceil((double) i/mid);
        }
        return total<=h;
    }
    public int minEatingSpeed(int[] piles, int h) {
        int low = 1, high = sum(piles);
        while(low<=high){
            int mid = (int) ((low+high)/2);
            if(isPossible(mid,piles,h)){
                high = mid-1;
            }else{
                low = mid+1;
            }
        }
        return low;
    }
}

File: 907-koko-eating-bananas/koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        def isPossible(k):
            cnt = 0
            for pile in piles:
                cnt += (math.ceil(pile/k))
            return cnt<=h

        
        low = 1
        high = max(piles)
        ans = high
        while low<high:
            mid = (low+high)//2
            if isPossible(mid):
                ans = mid
                high = mid
            else:
                low = mid+1
        
        return ans